[dfsan] Add dataflow sanitizer support for SystemZ

anoopkg6 opened the pull request:
Add SystemZ specific changes for dataflow sanitizer on top of following two common code changes
i) Fix Endianness issue [#162881](https://github.com/llvm/llvm-project/pull/162881)
ii) Fix ShadowAddress computation [#162864](https://github.com/llvm/llvm-project/pull/162864[](url))

See conversation in original pr#[#162195](https://github.com/llvm/llvm-project/pull/162195)

Commit:
[dfsan] Add dataflow sanitizer support for SystemZ on top of #162881 and #162864(Refer original pr#162195)

thurstond commented on compiler-rt/test/dfsan/pair.cpp:
```diff
@@ -66,8 +66,10 @@ void test_simple_constructors() {
 #ifdef O0
-  assert(dfsan_read_label(&i1, sizeof(i1)) == 10);
-  assert(dfsan_read_label(&ptr1, sizeof(ptr1)) == 10);
+  assert(dfsan_read_label(&i1, sizeof(i1)) == 8 ||
```
IIUC the test changes are only needed for SystemZ, and therefore make the tests unnecessarily weaker for other platforms. Is it possible to make the alternative options only allowed for SystemZ?

anoopkg6 replied:
SystemZ identifies labels for struct elements precisely across all optimizations consistently because of the way struct return is handled.

anoopkg6 committed:
Merge branch 'llvm:main' into dfsan_systemz

Commit:
Modify struct copy tests to accept union of labels in struct  and individual label at -O0 for SystemZ only

thurstond reviewed (approved):
Thanks, looks good to me! Please wait a few days in case @browneee has any comments or concerns.

browneee commented on compiler-rt/test/dfsan/pair.cpp:
```diff
@@ -66,8 +66,15 @@ void test_simple_constructors() {
   int *ptr1 = pair1.first;
 
 #ifdef O0
+#  if !defined(__s390x__)
```
If we keep this #if/else, swap the logic:  `#if defined(__s390x__)` ...

anoopkg6 replied:
SystemZ identifies labels for struct elements precisely across all optimization levels.

browneee reviewed (commented):
Thanks for giving me the opportunity to comment. Sorry for delay due to being on vacation.

anoopkg6 committed:
Merge branch 'llvm:main' into dfsan_systemz

anoopkg6 committed:
Merge branch 'llvm:main' into dfsan_systemz

Commit:
1. Modify struct copy tests as SystemZ identifies labels for struct elements
   precisely across all optimization levels.
2. Other cleanup as suggested in code review.

anoopkg6 committed:
Merge branch 'llvm:main' into dfsan_systemz

Commit:
Merge branch 'dfsan_systemz' of github.com:anoopkg6/llvm-project into dfsan_systemz

anoopkg6 commented on compiler-rt/lib/dfsan/dfsan_allocator.cpp:
```diff
@@ -44,9 +44,24 @@ struct DFsanMapUnmapCallback {
+#if defined(__s390x__)
+const uptr kMaxAllowedMallocSize = 2UL << 30;  // 2G
+
+struct AP64 {  // Allocator64 parameters. Deliberately using a short name.
```
Done.

anoopkg6 commented on compiler-rt/test/dfsan/lit.cfg.py:
```diff
@@ -10,6 +10,8 @@
 # Setup default compiler flags used with -fsanitize=dataflow option.
 clang_dfsan_cflags = ["-fsanitize=dataflow"] + [config.target_cflags]
+if config.target_arch == "s390x":
+    clang_dfsan_cflags.append("-mbackchain")
```
Done.

browneee replied:
nit: Please phrase it a bit more explicitly:

`# s390x arch needs -mbackchain to print stack backtrace of origin.`

anoopkg6 commented on compiler-rt/test/dfsan/origin_endianness.c:
```diff
@@ -16,10 +16,10 @@ __attribute__((noinline)) FULL_TYPE foo(FULL_TYPE a, FULL_TYPE b) {
   FULL_TYPE a = 1;
   FULL_TYPE b = 10;
-  dfsan_set_label(4, (HALF_TYPE *)&a, sizeof(HALF_TYPE));
+  dfsan_set_label(4, (HALF_TYPE *)&a + 1, sizeof(HALF_TYPE));
```
I added this test for https://github.com/llvm/llvm-project/pull/162881. This was verified only on x86 before SystemZ support was added. This test has undefined behavior and will fail on SystemZ because of endianness issue.

Actually, this test is redundant and taken from already existing test origin_ldst.c in dfsan testsuite.

See converasations in https://github.com/llvm/llvm-project/pull/162881.

browneee reviewed (approved):
Thanks for the ping. Sorry you had to ping.

LGTM, with one suggestion to clarify a comment.

anoopkg6 committed:
Merge branch 'llvm:main' into dfsan_systemz

Commit:
Incorporated mbackchain comment feedback for better

anoopkg6 committed:
Merge branch 'llvm:main' into dfsan_systemz

Commit:
Merge branch 'dfsan_systemz' of github.com:anoopkg6/llvm-project into dfsan_systemz
